### ISSUE TITLE
MEMORY NETWORKS

### ISSUE THREAD

**Issue by Kosuke-Szk** [Kosuke-Szk/papers#4]
Jason Weston, Sumit Chopra & Antoine Bordes
Facebook AI Research
https://arxiv.org/abs/1410.3916
Submitted on 15 Oct 2014

# 概要
Memory Networksという新しいモデルを提案する。Memory Networksは長期記憶コンポーネントを用いて推論を行う。QAのタスクで性能を検証する。QAでは、長期記憶が動的な知識基盤として機能し、応答はテキストで行う。大規模QAと小規模QA両方で評価を行った。後半では、複数のサポート文章を連ねることで効果的に機能することを示す。

# イントロ
ほとんどの機械学習モデルには長期記憶の構成要素の部分部分を読んだり書いたりする手段や、その方法を推論とシームレスに組み合わせる方法が欠けている。それゆえに、機械学習モデルは昨今のコンピュータの潤沢なアセットを活かしきれていない。例えば、文章を与えてそれに対する応答を行うタスクを考えてみる。古きは、時刻tまでの単語系列情報でt+1の単語を予測するRNNモデルがこのタスクに対処できるはずだった。しかしながら、隠れ層や重みによってエンコードされた記憶は一般的に小さすぎるのと、過去からの事実を正確に記憶できるほど十分に区画化されていない。
本研究ではこのような問題をMemoryNetworksで解決することを目指す。

# Memory Networks
Memory Networksは記憶領域の _m_ と４つのコンポーネントI,G,O,Rで構成される。
#### I: Input feature map
入力を内部状態に変換する

#### G: generalization
新しい入力のもとで古い記憶を更新する

#### O: output feature map
新しい入力と現在のメモリ状態のもとで出力を形成する

#### R: response
出力を要求されるフォーマットに整形する

フローとして、I→G→O→Rの順に情報が流れる。
コンポーネント化されていることで、IGORにはそれぞれ機械学習手法をアプライすることが可能。

#### I component:
一般的なテキストの前処理を施せる。構文解析や照応解析など。

#### G component:
最も単純なGは入力I(x)をスロットの中に保存する形態。
H(x)がメモリのインデックスを選択するメソッドとして、他の場所に触れずに、GがH(x)の場所を更新することもできる。さらに洗練されたモデルとしては、新しい入力をもとに過去の記憶を遡り更新すべき記憶をupdateできるもの。

メモリが巨大な場合、メモリを構造化する必要がある。これはスロット選択メソッドH(x)で達成できる。もしそれが可能になると、Gは全てのメモリを扱う必要がなくなる。(H(x)によって選択されるサブセットのみ扱えば良い。)

メモリが満タンになった場合は、'忘却'することも可能。それぞれの記憶に有効性のスコアをつけておいて、スコアが低いものから削除していけばよい。

#### O and R component
O component は一般的にメモリから情報を読み取り推論する役割を持つ。例えば、どの記憶が良い回答をするために関連のあるものか判断するなど。そして最終的な回答Rを生成する。

# A MemNN Implementation for text
コンポーネントにNNが使われるMemory NetworksをMemNNと呼ぶことにする。
## Basic Model
文章を入力することを考える。この文章は次の利用可能なメモリにそのままの形で保存される。Gは新しいメモリに情報を保存するのみで、古い情報を更新することはない。
k 個(k =1,2) のサポートメモリで入力と記憶の関連性を測り情報検索を行う。
![2018-12-30 14 45 11](https://user-images.githubusercontent.com/18728946/50544672-9a247980-0c41-11e9-942f-292aae9b192c.png)
ふたつの場合は、一番目のイテレーションののちに２番目のサポートメモリを検索する。
![2018-12-30 14 45 17](https://user-images.githubusercontent.com/18728946/50544671-9a247980-0c41-11e9-8a83-f198aacffa06.png)
最終的なアウトプットoがmodule Rへ入力される。

最後に、Rはテキスト形式の出力をしなければならない。最も簡単なのは、m_ok をそのまま返してしまうことである。真に文章生成として機能するためにはRNNを生成モデルとして使用しなければならない。本研究では返答は単語一語で行われるという妥協点の仮説のもと以下のようなモデルを構築した。
![2018-12-30 14 53 34](https://user-images.githubusercontent.com/18728946/50544699-d4424b00-0c42-11e9-984d-a4312b4cc18f.png)
以下が応答の例
![2018-12-30 14 54 18](https://user-images.githubusercontent.com/18728946/50544702-fe940880-0c42-11e9-9df9-f6498e69104a.png)

![2018-12-30 14 57 18](https://user-images.githubusercontent.com/18728946/50544717-431fa400-0c43-11e9-8f8e-adf2ade5aa49.png)
![2018-12-30 14 56 49](https://user-images.githubusercontent.com/18728946/50544719-43b83a80-0c43-11e9-8558-fcbd48a90fff.png)
![2018-12-30 14 57 06](https://user-images.githubusercontent.com/18728946/50544718-43b83a80-0c43-11e9-838b-b94da7cc380e.png)
![2018-12-30 14 57 30](https://user-images.githubusercontent.com/18728946/50544715-431fa400-0c43-11e9-9558-2ebadd1fd0e5.png)

本研究ではスコア関数S_o, S_rは同じ形にした。
![2018-12-30 15 08 39](https://user-images.githubusercontent.com/18728946/50544812-e2916680-0c44-11e9-9984-c40f80aa3609.png)
U: n x D行列
Φ: オリジナルの文章をD次元の特徴量空間に変換する行列
D: number of features
n: embedding demension
特徴量の最もシンプルな選び方はBoW. 今回はS_oに対してD = 3|W|. それぞれの単語は3つの異なる表現を持つからである。1つはΦ_yの、2つはΦ_xのもの。ふたつというのは、入力からくる表現とサポートメモリからくる表現の二種。D = 3|W|はS_oとS_r共通で、U_rとU_oはそれぞれ違う重み行列。

### Training 
入力と出力を与えた教師あり学習を行う。
![2018-12-30 15 17 38](https://user-images.githubusercontent.com/18728946/50544861-66981e00-0c46-11e9-9bf7-4f62d1987d64.png)
γ: margin
f barやr barは負例. これらの関数を最小化する。

# Word sequences as input
入力が単語の系列で、かつ肯定文と疑問文に別れていない場合、セグメンテーションが必要になる。
![2018-12-30 15 26 23](https://user-images.githubusercontent.com/18728946/50544892-4ae14780-0c47-11e9-986b-52a20c660712.png)
cは入力BoW. W_segはemnedding表現を分類するためのパラメタ。seg(c) > γ (γ:margin)である場合、これはセグメントと認識される。この手法によってMemNNは学習コンポーネントを持つことができている。 論文中では最初のproof of conceptとされている。

#  EFFICIENT MEMORY VIA HASHING
メモリが巨大でスコア付が難しい場合、速度向上のためにhashing tricksというものを用いる。
Inputをいくつかのバケットに振り分ける。そしてInputは同じバケットの中身しかスコアリングを行わない。

# MODELING WRITE TIME
メモリにいつ書き込まれるか、を考慮にいれることでモデルを拡張できる。これは動かない事実に対するQAの際は不要である。(フランスの首都はどこか。など)
一方で、時系列的に事実が変化していくものに関しては重要。
これを実装する簡単な方法としては、書き込み時刻 _j_ の情報をΦ_j として埋め込むことである。
しかしこれは絶対時間を扱うことが要求され、ものごと間の相対時間を取り扱うことが不可能。
そこで、以下の式を導入する。
![2018-12-30 16 04 09](https://user-images.githubusercontent.com/18728946/50545083-964a2480-0c4c-11e9-9a33-30c4e1114148.png)
Φ_t(x,y,y') = 0 or 1. 
S_ot > 0の時、y over y'
S_ot < 0の時、y' over y

#  MODELING PREVIOUSLY UNSEEN WORDS
未知語が出てきた際に機械学習がどう対応すべきか。
ひとつの方法としては言語モデルを使うこと。意味の近しい単語であると仮定して、どの単語が当てられるべきかを予測する。そして未知語がその言葉に近いものであると推定する。

Bag of Wordsを用いる際、ある単語には左側の共起単語群と右側の共起単語群がある。これらの共起性から未知語を予測することが可能そうである。したがって、特徴量の次元を3|D|→5|D|に拡大する。
これらのモデルに対しては、'dropout'で学習させて対応する。

# EXACT MATCHES AND UNSEEN WORDS
Embeddingモデルは低次元nによる表現のため、正確なwordマッチングを行えない。よって以下の式が解決策として使える。
![2018-12-30 16 26 59](https://user-images.githubusercontent.com/18728946/50545223-45d4c600-0c50-11e9-8f8f-35a6f68b9382.png)
BoWマッチングスコアをEmbeddingスコアに付加する。
特徴量次元をD = 8|W|とする
 
# 関連研究
古典的なQAメソッドは文書のセットをメモリのように扱い、情報検索の手法で答えを見つけるものである。近年では知識ベース(KB)のグラフ構造を用いる手法が試されている。(クエリと回答を論理的に繋げる)
NNとEmbeddingをベースにした手法も近年研究されている。知識ベースの手法と比較して、メモリネットワークスは以下二つの段階を踏まないで良いところが差異である。
(i) 知識ベース(KB)を構築するために情報抽出を行う
(ii) KBに法って推論を行う

associative memory networksのような、古典的なNNメモリモデルはコンテンツにアクセス可能なメモリの提供を目指している。Keyベクトルをoutputベクトルへつなぐおのである。一般的にこのタイプのメモリは重みネットワークの全体に分布しており、区画化されているわけではない。一方で、メモリベース学習は全てのサンプルを検索するということはせず、最も近いラベルを探すのみである。Memory NetworksはNNモデルを区画化されたメモリネットワークを組み合わせることで、メモリに対する読み書きを習得できる。

ところが、すでに読み書き可能なメモリモデルを構築しているものがある。
・neural network pushdown automaton
push と popが実装されている。
・self referentially
Memory Addressingの一種
・DISCERN
・NARX

'RNNSearch'という機械翻訳の手法もある。長系列のインプットに対する性能の低さに対応すべく、alignment mechanismを用いる。

# 実験と結果
![2018-12-30 18 04 15](https://user-images.githubusercontent.com/18728946/50545715-58093100-0c5d-11e9-8f73-79b9d293713b.png)
#  LARGE-SCALE QA
QA dataset introduced in Fader et al. (2013)で実験。
“Who wrote the Winnie the Pooh books?” 
“Who is poohs creator?”.
などが質問として含まれる。

# SIMULATED WORLD QA
4 characters, 3 objects and 5 roomsの状況を作る。charactersはあちこち動き、ものを拾ったり落としたりする。このタスクによってストーリーの理解力を検証する。
![2018-12-30 18 10 42](https://user-images.githubusercontent.com/18728946/50545749-41170e80-0c5e-11e9-9929-68b9fcc2ac07.png)

![2018-12-30 18 12 25](https://user-images.githubusercontent.com/18728946/50545754-7cb1d880-0c5e-11e9-9aaf-36acfc092443.png)

![2018-12-30 18 13 25](https://user-images.githubusercontent.com/18728946/50545763-a0751e80-0c5e-11e9-9f8c-c33f6cb49193.png)

# QA WITH PREVIOUSLY UNSEEN WORDS
未知語への対応も観察した。ロードオブザリングのキャラクターに関する質問に、一度も見たことのない知識について回答できた（本当？）
MemNNはS V Oなどのシンプルな文構造をつかむことができる。

# COMBINING SIMULATED DATA AND LARGE-SCALE QA
LARGE-SCALE QAとSIMULATED DATAの両方で訓練したモデルで試験した。このモデルは質問に対して実世界の知識を使って回答することもできるし、ストーリーから得た知識を使って解くこともできる。
結果はFigure.4に示す
いくつかの回答は妥当だがそうでないものもある。将来的にはストーリーベースと実世界ベースの知識の両方をうまく組み合わせる手法を検討する必要がある。

# まとめ
QAタスクに対してMemory Networksを応用した。将来的にはopen domainのQAなどにも応用することが期待される。例えば、WebQuestionsなど、複数の論理推論が必要なものなど。
Gを洗練させることでさらに複雑な状況を扱えるようになる。
弱い教師あり学習も非常に重要。
Memory Networksは言語のみならず、CVなど多数の分野で応用されていくだろう。

# コメント
